MSC3773: Notifications for threads (#3773)
matrix-org/matrix-spec-proposals: proposals/3773-notifications-for-threads.md

# MSC3773: Notifications for threads

Since the unread notification count does not consider threads, a client is unable
to separate the unread message counts into threads (as defined by
[MSC3440](https://github.com/matrix-org/matrix-doc/pull/3440))
without iterating over every missing message. Without this, clients are unable to:

* Let users know that a thread has new messages since they last read it.
* Accurately display a count of unread messages in a room (or a thread).

## Proposal

### Modification to push rule processing

When an event which is part of a thread (i.e. has a valid `m.relates_to` with
`rel_type` of `m.thread`) matches a push rule which results in a `notify` action
then the homeserver should note that the notification applies in the context of

> the thread ID (parent event ID).

Reviewer: What do we mean by this? What should the server do? I think the two things are:

1. For `/sync` we need to store the notifications counts per-thread;
2. something to do with sending push to the gateways? Do we do something different if its for a thread?

Author: @erikjohnston This doesn't make any changes to what is sent to push gateways (that should be clarified). So it is pretty much only your first point -- storing the notification counts on a per-thread basis so that they can be returned as part of the `unread_thread_notifications` object in `/sync`.

Author: Hopefully 94bf9ac172e3e34cdf84bc6cca91655be8ce6132 makes this clear!

# MSC3773: Notifications for threads

Since the unread notification count does not consider threads, a client is unable
to separate the unread message counts into threads (as defined by
[MSC3440](https://github.com/matrix-org/matrix-doc/pull/3440))
without iterating over every missing message. Without this, clients are unable to:

* Let users know that a thread has new messages since they last read it.
* Accurately display a count of unread messages in a room (or a thread).

## Proposal

### Modification to push rule processing

When an event which is part of a thread (i.e. has a valid `m.relates_to` with
`rel_type` of `m.thread`) matches a push rule which results in a `notify` action
then the homeserver should partition the resulting notification count per-thread.
(This is needed for the [proposed `/sync` changes](#unread-thread-notifications-in-the-sync-response)).

Similar behavior should be applied for an event which results in `notify` action

> with a `highlight` tweak set.

Author: I'm not sure if it should be highlighted somewhere, but the thread root message would *not* get double counted as part of the thread also. I'm not sure if this is correct or not.

Reviewer 2: I think that makes sense. Everywhere in the spec, the root message is handled as "separate of the thread". The `/relations` endpoint returns a chunk of events and the original message is separate. 

I would only expect the root message to be counted as part of the main timeline and not the thread timeline

This MSC does not propoase any changes to the payload sent to push gateways.

### Unread thread notifications in the sync response

Threaded clients can opt into receiving unread thread notifications by passing
a new `unread_thread_notifications` parameter
[as part of the `RoomEventFilter`](https://spec.matrix.org/v1.2/client-server-api/#filtering).
(This is [similar to `lazy_load_members`](https://spec.matrix.org/v1.2/client-server-api/#lazy-loading-room-members),
but only applies to the `/sync` endpoint.):

* `unread_thread_notifications`:  If `true`, enables partitioning of unread notification
  counts by thread. Defaults to false.

> If this flag is set to `true`, for each ["Joined Room" in the `/sync` response](https://spec.matrix.org/latest/client-server-api/#get_matrixclientv3sync)

Reviewer 3: ```suggestion
If this flag is set to `true`, for each ["Joined Room" in the `/sync` response](https://spec.matrix.org/v1.3/client-server-api/#get_matrixclientv3sync)
```

"latest" means something different after this MSC is merged, so point to a time-locked version.

Author: Thanks! I thought I had caught them all, but maybe that was a different MSC. 👍

a new field is added:

* `unread_thread_notifications`: Counts of unread thread notifications for this
  room, an object which maps thread ID (the parent event ID) to
  `Unread Notification Counts`.

Additionally, the `unread_notifications` dictionary is modified to only include
unread notifications from events which are not part of a thread.

An example of a joined room from a sync response:

```json5
{
  "account_data": {
    // ...
  },
  "ephemeral": {
    // ...
  },
  "state": {
    // ...
  },
  "summary": {
    // ...
  },
  "timeline": {
    "events": [
      {
        "event_id": "$143273582443PhrSn:example.org",
        // other fields ...
      },
      {
        "event_id": "$SGNxGPGUopcPBUoTTL:example.org",
        "m.relates_to": {
          "event_id": "$143273582443PhrSn:example.org",
          "rel_type": "m.thread"
        }
        // other fields ...
      }
    ]
  },
  "unread_notifications": {
    "highlight_count": 2,
    "notification_count": 18
  },
  "unread_thread_notifications": {
    "$143273582443PhrSn:example.org": {
      "highlight_count": 0,
      "notification_count": 1
    }
  }
}
```

> ## Potential issues

Author: @gsouquet noted another potential issue with this proposal that I'm unsure of the exact details of -- it was something about removing thread IDs from `unread_thread_notifications` when the counts go to 0 and having to iterate the entire dictionary or something.

@gsouquet can you fill in details here of your concern?

Reviewer 2: I think I found a solution that would be adequate and work for most clients. You can see the implementation here, https://github.com/matrix-org/matrix-js-sdk/pull/2600/files#diff-1b0cb46cd1fcdaacf73c771bf7eed6e00a12ee030da1dbcca4ae12842e1ecd81R1303-R1323

The idea being to reset entirely the `threadNotifications` map rather than mutating the old values. I'd consider this problematic resolved

### Events related to thread events

Events which are related to thread events (e.g. a reaction to a message which is
part of a thread) would not be represented in the notification count for that
thread, it would still appear in the main timeline's notification count.

With the default push rules this does not seem to be a problem as thread notifications

> mostly apply to `m.room.message` and `m.room.encrypted` events.

Reviewer 4: Would it not be a problem when editing messages? Those would refer to an event in the thread, but still create a notification.

Author: I always assumed that edits to a message causing new (duplicate) notifications was a bug.

I think the alternative is to include some text like we did in MSC3771:

> It is recommended that at least 3 relations are traversed when attempting to find a thread, implementations should be careful to not infinitely recurse.

That would likely be doable if we're doing it there anyway.

Reviewer 4: > I always assumed that edits to a message causing new (duplicate) notifications was a bug.

Oh yes, it is. You're right in that I'm probably looking at the wrong end of the stick.

> I think the alternative is to include some text like we did in MSC3771:
> 
> > It is recommended that at least 3 relations are traversed when attempting to find a thread, implementations should be careful to not infinitely recurse.
> 
> That would likely be doable if we're doing it there anyway.

Hm, yes that would help. As you say, edit-based notifications are an invalid use case, but there may be others that aren't immediately apparent, which deeper traversal would enable.

So: not a blocker, but may be a nice-to-have if it's simple implementation-wise.

Author: I'll take a look, it would make push rules act a bit more reasonably.

Author: I updated this in 3f2d43508c474e16b6fe6f60e96d1b4fe4504448. It pretty much uses the same text as MSC3771.

### Scalability

Rooms with many unread threads could cause some downsides:

> * The size of the `/sync` response would increase without bound.

Reviewer 4: To help with this, could a homeserver choose to omit the `unread_thread_notifications` field altogether if the unread counts have not changed since the previous `/sync` response?

Author: Do you mean omit the entire field or the individual threads in it? If the latter, I think that would be quite hard to do, but might be doable! The former seems likely doable, but I don't love making it act differently than `unread_notifications`?

Reviewer 4: The former, omitting the field when a new event has appeared in the main timeline (and thus would not update unread thread counts. Especially since many rooms won't have threads in them.

> The former seems likely doable, but I don't love making it act differently than `unread_notifications`?

I agree, and we could change always make this change later. I also wouldn't mind making the same change to `unread_notifications` in the future.

* The effort to generate and process the receipts for each room would increase
  without bound.

This is not dissimilar to rooms which are never read, however, as their unread
counts are continually tracked and returned as part of the `/sync` response.

### Clearing unread notifications

This MSC does not attempt to modify how unread notifications (for a thread or
otherwise) are cleared. It currently assumes the rules set forth by

> [read receipts](https://spec.matrix.org/latest/client-server-api/#receiving-notifications)

Reviewer 3: ```suggestion
[read receipts](https://spec.matrix.org/v1.3/client-server-api/#receiving-notifications)
```

still apply. This will cause some flakiness with unread notifications, as the current
receipt infrastructure assumes that a room's timeline is linear, which is no
longer true.

[MSC3771](https://github.com/matrix-org/matrix-spec-proposals/pull/3771) is a

> potential solution for this.

Reviewer 2: Wondering whether this should touch on how unread notifications work or not.

Marking a room as read is currently fairly simple as one only needs to send a read receipt targetting the last room event and off you go.
With this new proposal, to mark a room as read, clients will need to send one read receipt per unread thread. This could generate a high volume of requests in large rooms

Author: This sounds like feedback on MSC3771, not on MSC3773?

Reviewer 2: I believe you addressed that in an offline conversation we were having.
One can clear an entire room counter by sending a read receipt without the `threadId` parameter, and therefore that read receipt will be applied to all timelines

## Alternatives

### Using push rules

It might seem that a new push rule `action` (or `tweak`) should be used to control
the behavior of whether an event generates a notification for a thread or the
room itself. There are issues with either approach though:

A new `action` (e.g. `notify_thread`) would mean that additional logic would
need to be defined and added for events which aren't part of a thread but attempt
to use this action. It also conflicts with [MSC3768](https://github.com/matrix-org/matrix-spec-proposals/pull/3768),
which attempts to define another `action` which should also work fine for threads.

> A new `tweak` (e.g. `threaded`) does make much sense since there is not need to

Reviewer 4: ```suggestion
A new `tweak` (e.g. `threaded`) doesn't make much sense since there is no need to
```

?

Author: Ugh, thanks! I'm choosing slightly different text below to avoid the double negative.

```suggestion
A new `tweak` (e.g. `threaded`) was discard as an option since since there is no need to
```

* Let users know that a thread has new messages since they last read it.
* Accurately display a count of unread messages in a room.

To illustrate the issue with the current unread notification counts, we reuse
the example from [MSC3771](https://github.com/matrix-org/matrix-spec-proposals/pull/3771):

```mermaid
flowchart RL
    subgraph Main timeline
    C-->B
    B-->A
    F-->C
    end
    subgraph Thread timeline
    D-->A
    E-->D
    end

    style A fill:cyan,stroke:#333,stroke-width:2px
    style E fill:orange,stroke:#333,stroke-width:2px
```

While viewing the "main" timeline of the room, a client might move the read
receipt from event `A` to event `E`. While the read receipt is at `A`, the unread
notifications would be 5, once it moves to `E` it would be 1. Neither of these
numbers make much sense to the user:

* They have only read 3 messages (`B`, `C`, and `E`), but the unread count dropped by 4.
* There's no more messages to read, but it reports 1 unread message.

## Proposal

### A new push rule action: `relation`

In order for clients to separate the notification of threads from the main timeline,
a new `action` is proposed for push rules:

> * `notify_thread`: This causes each matching event to generate a notification

Reviewer 5: Probably a stupid question but if this is meant to be used together with `related_event_match` and `relation_match`, couldn't the server automatically infer this treatment without a dedicated `notify_thread` action?

Author: @t3chguy pretty much summed up the reasoning. I did think about whether it should happen automatically, but that didn't seem to match the way push rules are designed.

Reviewer 5: What confuses me is that this allows to combine `notify_thread` with conditions that match events _outside_ a thread. In fact, `notify_thread` seems to be both an action and a condition.

> If the event is not part of a thread the behavior is the same as `dont_notify`.

Author: It seemed difficult to concretely separate that -- what does `notify_thread` mean if the event wasn't part of a thread at all? Yet making `notify` implicitly have different behavior for events in threads doesn't seem correct either. If you have an idea of how to break this connection more that sounds interesting!

Reviewer 5: I bring no solutions, only problems, sorry. 😅

But essentially `notify_thread` does the same thing as `notify` and in addition also increments a thread-specific counter, right? Having `notify` do that directly depending on whether the event is part of a thread seems pretty natural to me personally.

When combined with [MSC3768](https://github.com/matrix-org/matrix-spec-proposals/pull/3768), not having `notify_thread` would also avoid the situation where [`notify_thread`] alone implies `notify` but [`notifiy_in_app`, `notify_thread`] doesn't.

Author: Thinking through this more I think the way to do this is the implicit logic; I think it gives reasonable flexibility in that you don't need to potentially provide rules many times *and* it breaks the requirement of MSC3772.

Author: I ended up heavily simplifying this in b4755fcd2e2a5cc87b72188e571a4cb3251b3882, so I'm going to close this thread, but we can reopen it if needed.

  in the context of a thread matching the `m.relates_to` information. If the event
  is not part of a thread the behavior is the same as `dont_notify`.

[Similarly to the `notify` action](https://spec.matrix.org/v1.2/client-server-api/#receiving-notifications),
this increments the notification count for a thread specific notification count
n addition to the overall room's unread notification count.  Setting a `highlight`
tweak will cause the highlight count to increase in addition to the notification

> count.[^1]

Reviewer 6: bundling in highlight feels a bit magical without explicitly saying highlight_thread, and what if you wanted `highlight` without `notify` but still be aggregated by thread, maybe it should be a modifier so that you need all three, `threads_notify, notify, highlight` where then `threads_notify` alone actually wouldn't do anything, alternatively maybe `set_tweak counter = threads` or something

Author: > what if you wanted `highlight` without `notify` but still be aggregated by thread

`highlight` is a `tweak` of `notify` and implies `notify` (you cannot `highlight` without `notify`). I can't seem to find a spec reference to this, but it was pretty much that the "notification count" must be <= the "highlight count".

> maybe it should be a modifier so that you need all three, `threads_notify, notify, highlight` where then `threads_notify` alone actually wouldn't do anything, alternatively maybe `set_tweak counter = threads` or something

I think this is suggesting that instead of a `notify_thread` `action` it should be a `tweak`? There's a section in the alternatives about this: https://github.com/matrix-org/matrix-spec-proposals/blob/clokep/threads-notifications/proposals/3773-notifications-for-threads.md#an-action-instead-of-a-tweak

I had originally written this MSC as it being a `tweak` and not a separate `action`, but it seemed like a poor fit since the push server likely wouldn't care about it.

An example of a joined room from a sync response:

```json5
{
  "account_data": {
    // ...
  },
  "ephemeral": {
    // ...
  },
  "state": {
    // ...
  },
  "summary": {
    // ...
  },
  "timeline": {
    "events": [
      {
        "event_id": "$143273582443PhrSn:example.org",
        // other fields ...
      },
      {
        "event_id": "$SGNxGPGUopcPBUoTTL:example.org",
        "m.relates_to": {
          "event_id": "$143273582443PhrSn:example.org",
          "rel_type": "m.thread"
        }
        // other fields ...
      }
    ]
  },
  "unread_notifications": {
    "highlight_count": 2,
    "notification_count": 18
  },
  "unread_threads": {
    "$143273582443PhrSn:example.org": {
      "highlight_count": 0,
      "notification_count": 1
    }
  }
}
```

Similarly to read receipts, when the user updates their `m.read.thread.private`
read receipt from [MSC3771](https://github.com/matrix-org/matrix-spec-proposals/pull/3771)
(either by using the API or by sending an event into the thread), notifications
prior to and including that event in the thread MUST be marked as read. In order
to keep the unread counts for the room, the equivalent amount of unread

> notifications for the room are decremented at the same time.

Reviewer 6: My worry with this is that as per the threads MSC, [clients don't need full threads support](https://github.com/matrix-org/matrix-spec-proposals/blob/gsouquet/threading-via-relations/proposals/3440-threading-via-relations.md#sending-mthread-before-fully-implementing-threads) - so may send normal RRs for thread-responses, possibly causing that user's thread's capable clients to do the math of `unread_notifications. notification_count - sum(unread_threads...notification_count)` and get a negative number

As well as causing the threads to not be marked as read (wrongly)

Author: I had originally written this as not changing the room unread counts for backwards compatibility, but then convinced myself it wasn't needed since an old client would never send a thread read receipt. I think I failed to consider mixed thread-aware and non-thread-aware clients though.

Will have another think about that! 👍

Author: How exactly this behaves probably depends slightly on an open thread from MSC3771: https://github.com/matrix-org/matrix-spec-proposals/pull/3771#discussion_r857810047

I'm going to park this until we have thoughts for that.

Author: I've punted this back to MSC3771.

## Potential issues

Trying to support backwards compatibility with the unread notifications count may
prove finicky, but should be workable. This creates a significant amount more
bookkeeping for the homeserver which could prove to be expensive.

This has similar caveats to MSC3771 about the size of the `/sync` response -- if
there are threads which are never read it the notification counts will continually
appear in the sync response. This is not dissimilar to rooms which are never read,
however, as their unread counts are tracked and returned.

This is somewhat incompatible with [MSC3768](https://github.com/matrix-org/matrix-spec-proposals/pull/3768)
as that adds a new action as well (`notify_in_app`). It should be possible to
layer in support without having to create a `notify_thread_in_app` action, but

> this hasn't been fully considered:

Reviewer 6: My set_tweak suggestion may sidestep this

| `actions`                        | Implied value                                          |
|----------------------------------|--------------------------------------------------------|
| [`notify`]                         | `notify`                                               |
| [`notify_thread`]                  | `notify` + `notify_thread` (MSC3773 behavior)          |
| [`notify_in_app`]                  | `notify_in_app`                                        |
| [`notify_in_app`, `notify_thread`] | `notify_in_app`, `notify_thread` (MSC3768 for threads) |

> ## Alternatives

Reviewer 5: For whatever it's worth, Slack doesn't display thread notifications on the room the thread is in but instead on a _single_ "Threads" item in the room / contact list.

<img width="700" alt="threads" src="https://user-images.githubusercontent.com/1137962/164192559-2a453368-e9c3-403e-bf7e-85f2a72ac4a0.png">

This is obviously a lot more limited than what's proposed in this MSC. At the same time, it could make for an easier to implement alternative because this way you don't have to mix and maintain room and thread notification counts.

Author: I'm failing to grok how that's related to this MSC -- it seems more related to MSC3440 which defines how threads exist (and essentially includes them as part of the room already).

Reviewer 5: Sorry, I was trying to say that for simpler client UIs, combining room and thread unread counts might not be required but I guess the spec needs to offer more flexibility than just the minimum so I'm withdrawing my comment.
